### PR TITLE
27 internationalization i18n

### PR DESCRIPTION
**Pull request by HannaRosberg**
**Description**
This pull request introduces internationalization (i18n) support to the frontend, allowing the application to be displayed in three languages:

- English (default)
- Swedish
- Spanish

**How to Test**
- Run the frontend locally:
  - npm install
  - npm start

**Key Changes**

**1. Added useTranslationStatus Hook**
-Created a custom React hook (useTranslationStatus) located in hooks/useTranslationStatus.tsx.
-The hook tracks the translation loading state (loading) and handles errors (error) during the translation process.
-The hook utilizes the i18next library to listen for translation-related events (loading, loaded, failedLoading).

**2. Integration of Translation Status in Components
Header Component**
-Added the useTranslationStatus hook to display loading or error messages when translation data is not yet available.
-Updated the navigation menu items to dynamically use translated keys (homekey, aboutkey, contactkey, loginkey) via the useTranslation hook.
-Modified the useEffect hook to log translation changes for debugging.

**Footer Component**
-Similar to the Header, integrated the useTranslationStatus hook to handle loading and error states.
-Replaced static text with translated keys (navkey, mediakey, addresskey, phonekey, emailkey).
-Dynamically rendered menu items and contact information using translation keys.

**LanguageSelector Component**
-Added a language selector dropdown that allows users to switch between supported languages (English, Español, Svenska).
-Utilized the i18n.changeLanguage() method to handle language changes.
-Logged language change actions for debugging purposes.

**TranslationHandler Component**
-Created a small utility component to display the translation loading state or any errors, using the useTranslationStatus hook.

**3. App Component Enhancements**
-Leveraged the useTranslation hook for dynamically displaying text based on the selected language.
-Introduced a key state to force component re-rendering upon language changes.
-Added loading and error handling for translations using useTranslationStatus.

**4. i18n Configuration Setup**
-Configured i18next to load translations using the Backend plugin with HTTP requests.
-Added LanguageDetector to detect the user's preferred language through various methods (e.g., query string, cookie, local storage, navigator).
-Enabled fallback to English (fallbackLng: 'en') if the user's language is unsupported.
-Configured detection options to prioritize language determination.

**5. UI/UX Improvements**
-Improved the responsiveness of the Header by toggling mobile menu visibility based on user actions.
-Updated the navigation menu to utilize dynamic translations.
-Styled loading and error messages to ensure better user feedback during translation initialization or errors.

**Verify that:**
-The app defaults to English.
-Language switching works as expected.
-All translated text changes dynamically.

### **Why we need the multilingual features**
This PR is part of the multilingual feature initiative to support users in different regions, aligning with the project's goal to improve accessibility and user experience.

image

image

image

## Release Notes

- **New Features**
  - Enhanced internationalization support with language selection capability.
  - Added new components: `LanguageSelector` and `TranslationHandler`.
  - Dynamic text rendering in the `Header` and `Footer` components.

- **Translations**
  - Introduced English, Spanish, and Swedish translation files with key UI elements and festive messages.

- **Bug Fixes**
  - Improved error handling and loading states for translations.

- **Styling**
  - Added new CSS classes for improved styling of error and loading messages.

### CODE REVIEW

**Review by ConnyPettersson — DISMISSED** [review]
Overall, the code appears clean and easy to follow, especially in App.tsx where the language buttons and text strings are clearly separated. If additional translatable content or languages are introduced, it could be beneficial to extract the language-switching logic and a possible button generator into a separate component for better separation of logic and presentation. Adding tests to ensure long-term quality would also be worth considering.

**Review by MikaelaMathisson — CHANGES_REQUESTED** [review]
This PR solves the issue for Internationalization well.

I think the code is simple. very readable and has great variable names. Its easy to follow.

I did notice though, that the coderabbit suggested the following:
 "Add error handling and loading state for translations"
I think this is necessary for this type of issue and would recommend looking through the changes that the coderabbit suggested, to make the code handle error or loading states.

File: frontend/src/i18n.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        },
+        es: {
+            translation: {
+                "welcome": "Bienvenido a React",
+                "changeLanguage": "Cambiar idioma",
+                "visitCake": "Visitar /cake",
+                "cakeText": "¡Disfruta de una deliciosa tarta!"
+            },
+        },
+    },
```
[the remark this answers is not in the document]
Comment (MatsJoh50) [reply frontend/src/i18n.ts]: As one who never worked with i18n I feel that this is a great way to keep each page clean and easy to work with, especilly when the app grows with more propper content.

**Review by MatsJoh50 — DISMISSED** [review]
Great start for i18n and using /cake as a small tutorial/sample is  great!

I like the suggestion from coderabbit to  refactor each page to a seperate translation file for ease of scaling.

**Review by MikaelaMathisson — APPROVED** [review]
I think this looks great after the changes. Good job on adding error and loading states 👍

**Review by oskber — APPROVED** [review]
Nice work with the translation implementation 👍

**Pull request merged**